File: middleware.rb/resources/attempt.rb
require 'open-uri'

  def shellescape(str)
    # An empty argument will be skipped, so return empty quotes.
    return "''" if str.empty?

    str = str.dup

    # Process as a single byte sequence because not all shell
    # implementations are multibyte aware.
    str.gsub!(/([^A-Za-z0-9_\-.,:\/@\n])/n, "\\\\\\1")

    # A LF cannot be escaped with a backslash because a backslash + LF
    # combo is regarded as line continuation and simply ignored.
    str.gsub!(/\n/, "'\n'")

    return str
  end

module Middleware
  
  class Application < Sinatra::Base
    
    post '/attempt' do
      # pre-check: user may start test? user has test running?
      # pre-check: user allowed to attempt?
      # pre-check: test allows attempt/mode
      
      # Resume running attempt for this test?
      
      ea = db['attempts'].find_one({
        'user_id' => BSON::ObjectId(@params[:user_id]),
        'test_id' => BSON::ObjectId(@params[:test_id]), 
        'state' => { '$ne' => 'CLOSED' }
      })
      
      if ea
        status 303 # See Other
        headers \
          'Location' => '/attempt/'+ea['_id'].to_s
        return json({'_id'=>ea['_id'].to_s})
      end
      
      # User has another attempt running?

      eua = db['attempts'].find_one({
        'user_id' => BSON::ObjectId(@params[:user_id]),
        'state' => { '$ne' => 'CLOSED' }        
      })
      
      if eua
        status 403 # Forbidden
        return json({'msg' => 'User already has a running attempt.'})
      end
      
      # Check preconditions, then safe to create a new attempt
      
      t = db['tests'].find_one({'_id' => BSON::ObjectId(@params[:test_id])})
      
      unless t
        status 404
        return json({"msg"=>"Test does not exist"})
      end

      a = []
      
      t['content'].each do |remotetest|
        service = db.dereference(remotetest['service'])

        url = service['url'] + "/attempt"

        response = RestClient.post(url, {
          :user_id => 0,
          :test_id => remotetest['test']
        })
        
        r = JSON.parse(response.body)
        
        a << r['id']
      end


      aid = db['attempts'].insert({
        'attempts' => a,
        'user_id' => BSON::ObjectId(@params[:user_id]),
        'test_id' => BSON::ObjectId(@params[:test_id]),
        'test' => t,
        'state' => 'CREATED',
        'cursor' => 1,
        'date' => '',
        'form_url' => @params[:form_url],
        'asset_url' => @params[:asset_url]
      })
      
      status 201 # Created
      headers 'Location' => '/attempt/'+(aid.to_s)
      json({"_id" => aid.to_s})
    end
              
    get '/attempt/:id/page/:n' do
      check_id!
      
      a = db['attempts'].find_one({'_id' => BSON::ObjectId(@params['id'])})
      raise Sinatra::NotFound unless a
      
      page = a['test']['content'][@params[:n].to_i-1]
      attempt = a['attempts'][@params[:n].to_i-1]
      
      #determine service
      service = db.dereference(page['service'])
      
      #determine url
      url = service['url'] + "/attempt/#{attempt}/page/#{page['page']}"
      
      response = RestClient.get(url)
      r = JSON.parse(response.force_encoding("UTF-8").gsub("\xEF\xBB\xBF".force_encoding("UTF-8"), ''))
      
      asset_base = a['asset_url'] % { :service => service['_id'] }
      
      o, s = Open3.capture2("./lib/assetize/assetize.js #{shellescape(asset_base)}", :stdin_data=>r['content'])

      #fetch and deliver data
      o
    end
    

    
    
    get '/attempt/:id' do
      check_id!
      a = db['attempts'].find_one({'_id' => BSON::ObjectId(@params['id'])})
      
      raise Sinatra::NotFound unless a

      # a['cursor'] = 1
      a['state']  = 'READY'
      
      json a
    end
    
    put '/attempt/:id' do
      check_id!
      
      id = @params.delete('id')
      #@params.delete('commit')
      db['attempts'].update({'_id'=>BSON::ObjectId(id)}, {"$set" => {
        :cursor => @params[:cursor].to_i
      }})
      
      "{}"
    end
    
    put '/attempt/:id/page/:cursor' do
      check_id!
      
      id = @params.delete('id')
      cursor = @params.delete('cursor')
      a = db['attempts'].find_one({'_id' => BSON::ObjectId(id)})
      
      raise Sinatra::NotFound unless a
      
      page = a['test']['content'][cursor.to_i-1]
      attempt = a['attempts'][cursor.to_i-1]
      
      #determine service
      service = db.dereference(page['service'])
      
      #determine url
      url = service['url'] + "/attempt/#{attempt}/page/#{page['page']}"
      
      response = RestClient.post(url, @params)
      
      # save status from response to database...
      r = JSON.parse(response.body.gsub!("\xEF\xBB\xBF".force_encoding("UTF-8"), ''))
      
      
      
      # then redirect to feedback (optional: if feedback is enabled)
      
    end
    
    
    #   server.put(null, '/attempt/:id', pre, function(req, res, next) {
    #     delete req.params.id;
    #     Attempt.update({ _id: req.uriParams.id }, { $set: req.params }, function(err) {
    #       if (err) {
    #         res.send(500, err);
    #       } else {
    #         res.send(200, {});
    #       }
    #       return next();
    #     });
    #   }, post);
    # 
    #   server.get(null, '/attempt', pre, function(req, res, next){
    # 
    #   });
    # }
    #     
  end

end

File: middleware/api/attempt.js
var mongoose = require('mongoose');

var async = require('async');

var Attempt = mongoose.model('Attempt');
var Test = mongoose.model('Test');

exports.setupRoutes = function (server, pre, post) {
  server.post(null, '/attempt', pre, function(req, res, next) {
    
    // // pre-check: user may start test? user has test running?
    //     async.parallel([
    //       
    //       // pre-check: another attempt running?
    //       function(callback) {
    //         Attempt.find({'user': req.params.user}, ['_id'], function(err, docs) {
    //           if (err) {
    //             res.send(500, err);
    //           } else {
    //             res.send(200, docs);
    //           }
    //           return next();
    //         });
    //         
    //       },
    //       
    //       // pre-check: user allowed to attempt?
    //       function(callback) {
    //         
    //       },
    //       
    //       // pre-check: test allows attempt/mode
    //       
    //     ], function(err, res) {
    //       
    //     });
    //         
    //     var a = new Attempt({
    //       title: req.params.title,
    //       date: Date.now()
    //     });
    //     
    //     t.save(function(err){
    //       res.send(500, err);
    //       return next();
    //     });
    
    var isnew;
    
    // Attempt.find({'user': req.params.user_id, 'test'}, ['_id'], function(err, docs) {
    //           if (err) {
    //             res.send(500, err);
    //           } else {
    //             res.send(200, docs);
    //           }
    //           return next();
    //         });
    // 
    // 
    // res.send(isnew ? 201 : 303, '', {'Location': '/attempt/'+t._id}); // created / see other
    
    Attempt.findOne({'user': req.params.user_id, 'test_id': req.params.test_id, 'state': { $ne : 'CLOSED' }}, ['_id'], function(err, doc) {
      
      if (err) {
        res.send(500, err);
      } else {
        
        if (doc) {
          console.log('sending 303 because attempt already open');
          res.send(303, {_id:doc._id}, {'Location': '/attempt/'+doc._id});
          return next();
          
        } else {
          
          // another attempt open for user?
          Attempt.findOne({'user': req.params.user_id, 'state': { $ne : 'CLOSED' }}, ['_id'], function(err, doc) {
            
            if (err) {
              res.send(500, err);
              return next();
            } else {
              
              if (doc) {
                
                res.send(403, {msg: "User already has a running attempt."});
                return next();
                
              } else {
                
                // no existing or running attempts found... safe to create a new one...
                // IF!! user is allowed to and test allows it... (TODO)
                
                Test.findById(req.params.test_id, function(err, tdoc) {
                  if (err) {
                    res.send(500, err);
                    return next();
                  }
                  
                  if (tdoc) {
                    var a = new Attempt({
                      user: req.params.user_id,
                      test_id: req.params.test_id,
                      state: 'CREATED',
                      date: Date.now(),
                      form_url: req.params.form_url,
                      asset_url: req.params.asset_url
                    });
                    
                    a.test.push(tdoc);

                    a.save(function(err){
                      res.send(500, err);
                      return next();
                    });

                    res.send(201, {_id:a._id}, {'Location': '/attempt/'+a._id});
                    return next();
                    
                  } else {
                    res.send(404, {msg: 'Test does not exist'});
                    return next();
                  }
                });
                
              }
              
            }
            
          });          
        }
        
      }
    
    
    });
    
    
  }, post);
  
  server.get(null, '/attempt/:id/page/:num', pre, function(req, res, next) {
    var jsdom = require('jsdom');

    jsdom.env("http://localhost:9006/test/1/page/1", [
      'http://code.jquery.com/jquery.min.js'
    ],
    function(errors, window) {
      if (errors) {
        console.log('error fetching remote resource');
        res.send(500, {
          msg: 'error fetching remote resource'
        });
        return next();
      }

      var css = [], js = [];
      window.$('link[rel=stylesheet]').each(function() { css.push(this.href); });
      window.$('script[src]').each(function() { js.push(this.src); });

      res.send(200, {
        javascripts: js,
        stylesheets: css,
        content: window.$('body').html()
      });
      return next();

    });
  });
  
  server.get(null, '/attempt/:id', pre, function(req, res, next) {
    Attempt.findById(req.uriParams.id, function(err, doc){
      if (err) {
        res.send(500, err);
        return next();
      } else {
        if (doc) {
          var a = {
            _id: doc._id,
            state: 'READY',
            form_url: doc.form_url,
            asset_url: doc.asset_url,
            user: doc.user,
            test: doc.test,
            cursor: 1
          };
          res.send(200, a);
        } else {
          res.send(404, {});
        }
      }
    });
  });
  
  server.put(null, '/attempt/:id', pre, function(req, res, next) {
    delete req.params.id;
    Attempt.update({ _id: req.uriParams.id }, { $set: req.params }, function(err) {
      if (err) {
        res.send(500, err);
      } else {
        res.send(200, {});
      }
      return next();
    });
  }, post);

  server.get(null, '/attempt', pre, function(req, res, next){
    
  });
}